src/typed-geometry/functions/objects/intersection.hh

```diff
@@ -2960,4 +2975,184 @@ template <class ScalarT>
     return intersects(box, hs);
 }
 
+// segment3 - halfspace3
+template <class ScalarT>
+[[nodiscard]] constexpr bool intersects(segment<3, ScalarT> const& s, halfspace<3, ScalarT> const& hs)
+{
+    if ((dot(hs.normal, s.pos0) - hs.dis) <= 0 || (dot(hs.normal, s.pos1) - hs.dis) <= 0)
+        return true;
+
+    return false;
+}
+
+template <class ScalarT>
+[[nodiscard]] constexpr bool intersects(halfspace<3, ScalarT> const& hs, segment<3, ScalarT> const& s)
+{
+    return intersects(s, hs);
+}
+
+// triangle3 - sphere3
+template <class ScalarT>
+[[nodiscard]] constexpr bool intersects(triangle<3, ScalarT> const& t, sphere<3, ScalarT> const& s)
+{
+    // triangle vertex inside sphere
+    if (contains(s, t.pos0) || contains(s, t.pos1) || contains(s, t.pos2))
+        return true;
+
+    plane<3, ScalarT> plane_t = plane_of(t);
+
+    // check if the closest point on triangle to sphere center is inside the sphere
+    auto cp = closest_points(s.center, t);
+
+    if (contains(s, cp.first) && contains(s, cp.second))
+        return true;
+
+    // triangle edge intersects sphere
+    for (auto const& e : edges_of(t))
+    {
+        if (intersects(e, s))
+            return true;
+    }
+
+    return false;
+}
+
+template <class ScalarT>
+[[nodiscard]] constexpr bool intersects(sphere<3, ScalarT> const& s, triangle<3, ScalarT> const& t)
+{
+    return intersects(t, s);
+}
+
+// sphere2in3 - plane3
+template <class ScalarT>
+[[nodiscard]] constexpr bool intersects(sphere<2, ScalarT, 3> const& s, plane<3, ScalarT> const& p)
+{
+    auto plane_s = tg::plane3(s.normal, s.center);
+
+    // sphere center on plane
+    if (contains(p, s.center))
+        return true;
+
+    // no intersection if planes are parallel
+    if (plane_s.normal == p.normal && !contains(p, s.center))
```

Comment:
I think we have a `are_parallel` function. the current test would not detect normal == -normal for example

Reply:
could not find a _are_parallel_ function in typed_geometry. should I then just add the case normal == -normal?